votor/src/timer_manager/timers.rs

```diff
@@ -0,0 +1,243 @@
+use {
+    crate::event::VotorEvent,
+    crossbeam_channel::Sender,
+    solana_ledger::leader_schedule_utils::last_of_consecutive_leader_slots,
+    solana_sdk::clock::Slot,
+    std::{
+        cmp::Reverse,
+        collections::{BinaryHeap, HashMap},
+        time::{Duration, Instant},
+    },
+};
+
+/// Encodes a basic state machine of the different stages involved in handling
+/// timeouts for a window of slots.
+enum TimerState {
+    /// Waiting for the DELTA_TIMEOUT stage.
+    WaitDeltaTimeout {
+        /// The slots in the window.  Must not be empty.
+        window: Vec<Slot>,
+        /// Time when this stage will end.
+        timeout: Instant,
+    },
+    /// Waiting for the DELTA_BLOCK stage.
+    WaitDeltaBlock {
+        /// The slots in the window.  Must not be empty.
+        window: Vec<Slot>,
+        /// Time when this stage will end.
+        timeout: Instant,
+    },
+    /// The state machine is done.
+    Done,
+}
+
+impl TimerState {
+    /// Creates a new instance of the state machine.
+    ///
+    /// Also returns the next time the timer should fire.
+    fn new(slot: Slot, delta_timeout: Duration) -> (Self, Instant) {
+        let mut window = (slot..=last_of_consecutive_leader_slots(slot)).collect::<Vec<_>>();
+        assert!(!window.is_empty());
+        window.reverse();
```

Comment:
Would it make more sense to use a VecDeque and pop_front so we don't have to store the window backwards?